### PR TITLE
Add `@pattern` support for constraint package 

### LINKED ISSUE #3179: Add `@pattern` support for constraint package
This was not implemented due to no support for Regex in the language.

https://github.com/ballerina-platform/ballerina-standard-library/issues/2850

### PR DESCRIPTION
**Pull request by TharmiganK**
## Purpose

> $Subject

Fixes : [`Add @pattern support for constraint package #3179`](https://github.com/ballerina-platform/ballerina-standard-library/issues/3179)

## Examples

```ballerina
import ballerina/log;
import ballerina/constraint;

string:RegExp regExp = re `([0-9]{10})|(\+[0-9]{11})`;

@constraint:String {pattern: regExp}
type PhoneNumber string;

type User record {|
    string name;
    @constraint:String {pattern: re `male|female`}
    string gender?;
    int age;
    @constraint:String {pattern: re `([0-9]{9}[v|V]|[0-9]{12})`}
    string nic;
|};

type UserAdvanced record {|
    *User;
    PhoneNumber contactNumber;
    @constraint:String {pattern: re `([a-zA-Z0-9._%\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,6})*`}
    string email;
|};

public function main() {
    PhoneNumber phoneNumber = "0774563957";
    do {
        phoneNumber = check constraint:validate(phoneNumber);
        log:printInfo("phone number is valid");
    } on fail error err {
        log:printError("error occurred while checking phone number", err);
    }

    User user = {
        name: "John",
        age: 36,
        gender: "male",
        nic: "198654673682"
    };
    do {
        user = check constraint:validate(user);
        log:printInfo("user is valid");
    } on fail error err {
        log:printError("error occurred while checking user", err);
    }

    UserAdvanced userAdvanced = {
        ... user,
        contactNumber: phoneNumber,
        email: "john@wso2.com"
    };
    do {
        userAdvanced = check constraint:validate(userAdvanced);
        log:printInfo("userAdvanced is valid");
    } on fail error err {
        log:printError("error occurred while checking userAdvanced", err);
    }
}
```

## Checklist
- [x] Linked to an issue
- [x] Updated the changelog
- [x] Added tests
- [x] Updated the spec
- [x] Checked native-image compatibility

### CODE REVIEW

**Comment by shafreenAnfar** [comment]
Didn't go through the code but the functionality looks good to me!

Hope @chamil321 will validate the tests and the code. Also lets make sure tooling side is ready when it comes to regex.

File: compiler-plugin/src/main/java/io/ballerina/stdlib/constraint/compiler/annotation/tag/LengthConstrainedAnnotationTag.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package io.ballerina.stdlib.constraint.compiler.annotation.tag;
```
Comment (chamil321) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/constraint/compiler/annotation/tag/LengthConstrainedAnnotationTag.java]: Missing licence header

File: compiler-plugin/src/main/java/io/ballerina/stdlib/constraint/compiler/annotation/tag/ValueConstrainedAnnotationTag.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package io.ballerina.stdlib.constraint.compiler.annotation.tag;
```
Comment (chamil321) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/constraint/compiler/annotation/tag/ValueConstrainedAnnotationTag.java]: Missing licence header

Comment (TharmiganK) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/constraint/compiler/annotation/tag/ValueConstrainedAnnotationTag.java]: Fixed via [5352f0c](https://github.com/ballerina-platform/module-ballerina-constraint/pull/29/commits/5352f0c82ae620f0f8610b86bf2d5bd50a897051)

File: compiler-plugin/src/main/java/io/ballerina/stdlib/constraint/compiler/annotation/tag/LengthConstrainedAnnotationTag.java (hunk 1 above)
Comment (TharmiganK) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/constraint/compiler/annotation/tag/LengthConstrainedAnnotationTag.java]: Fixed via [5352f0c](https://github.com/ballerina-platform/module-ballerina-constraint/pull/29/commits/5352f0c82ae620f0f8610b86bf2d5bd50a897051)

**Review by chamil321 — APPROVED** [review]

**Comment by TharmiganK** [comment]
> Didn't go through the code but the functionality looks good to me!
> 
> Hope @chamil321 will validate the tests and the code. Also lets make sure tooling side is ready when it comes to regex.

There are still some pending issues regarding the tooling support for regexp from lang side such as :
- https://github.com/ballerina-platform/ballerina-spec/issues/1188
- https://github.com/ballerina-platform/ballerina-lang/issues/38568
- https://github.com/ballerina-platform/ballerina-lang/issues/39249

**Pull request merged**